### PR TITLE
Fixng region and zone logic

### PR DESCRIPTION
**Pull request by 23TNC**
Region and zone logic needed to move to session.go because ssn requires both region and zone prior to platform.go's execution.  This could have also been moved to region.go however, session.go seems to be used to read the current env context so I put the functions there.

### CODE REVIEW

File: pkg/asset/installconfig/powervs/session.go [hunk 1]
```diff
@@ -21,8 +25,9 @@ var (
 
 // Session is an object representing a session for the IBM Power VS API.
 type Session struct {
-	Session *ibmpisession.IBMPISession
-	Creds   *UserCredentials
+	Session    *ibmpisession.IBMPISession
+	Creds      *UserCredentials
+	OsOverride string
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:30]: this override var doesn't really fit into the concept of a session. it's a purely openshift installer concept

File: pkg/asset/installconfig/powervs/session.go [hunk 2]
```diff
@@ -90,28 +102,161 @@ func getPISession() (*ibmpisession.IBMPISession, *UserCredentials, error) {
...
+	return uc, err
+}
+
+/*
+//  https://github.com/IBM-Cloud/power-go-client/blob/master/ibmpisession/ibmpowersession.go
+*/
+func getPISession(region string, zone string, uc *UserCredentials) (*ibmpisession.IBMPISession, error) {
+
+	// @TOOD: query if region is multi-zone? or just pass through err...
+	// @TODO: pass through debug?
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:120]: we can remove these TODO's now

File: pkg/asset/installconfig/powervs/session.go [hunk 3]
```diff
@@ -90,28 +102,161 @@ func getPISession() (*ibmpisession.IBMPISession, *UserCredentials, error) {
...
+	// Region
+	// -------------------------
+
+	// Retrieve sessionRegion from the current enviornment context
+	sessionRegion = os.Getenv("IBMCLOUD_REGION")
 	// this can also be pulled from  ~/bluemix/config.json
 	if r2 := os.Getenv("IC_REGION"); len(r2) > 0 {
 		if len(region) > 0 && region != r2 {
-			return nil, nil, errors.New(fmt.Sprintf("conflicting values for IBM Cloud Region: IBMCLOUD_REGION: %s and IC_REGION: %s", region, r2))
+			return "", "", errors.New(fmt.Sprintf("conflicting values for IBM Cloud Region: IBMCLOUD_REGION: %s and IC_REGION: %s", region, r2))
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:142]: i think you'd mentioned earlier that the code for the API keys doesn't check for conflicts? So it probably makes sense just to have one override other and not return an error here (to be consistent). IOW -- just delete this check.

File: pkg/asset/installconfig/powervs/session.go [hunk 4]
```diff
@@ -90,28 +102,161 @@ func getPISession() (*ibmpisession.IBMPISession, *UserCredentials, error) {
...
+		}, &region)
+		if err != nil {
+			return "", "", err
+		}
+	}
+
+	// -------------------------
+	// Zone
+	// -------------------------
+	// There is an effort to remove zone, I am uncertain how this can be done as it is required by
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:208]: delete this comment as we're not removing it

File: pkg/asset/installconfig/powervs/session.go [hunk 5]
```diff
@@ -90,28 +102,161 @@ func getPISession() (*ibmpisession.IBMPISession, *UserCredentials, error) {
...
+				return strings.SplitN(v, " ", 2)[0]
+			}
+			return ""
+		}
+
+		err = survey.Ask([]*survey.Question{
+			{
+				Prompt: &survey.Select{
+					Message: "Zone",
+					Help:    "The powervs zone within the region to be used for installation.",
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/session.go:241]: s/powervs/Power VS.
Also -- do all the survey questions help text have punctuation? This isn't a complete sentence so I wouldn't expect a period at the end.

File: pkg/asset/installconfig/powervs/platform.go [hunk 6]
```diff
@@ -1,132 +1,25 @@
...
-			}),
-			Transform: zoneTransform,
-		},
-	}, &zone)
-	if err != nil {
-		return nil, err
-	}
-
 	var p powervs.Platform
-	if osOverride := os.Getenv("OPENSHIFT_INSTALL_OS_IMAGE_OVERRIDE"); len(osOverride) != 0 {
```
Comment (clnperez) [inline pkg/asset/installconfig/powervs/platform.go]: this check should stay here. am i missing some technical reason why you moved it?

**Review by clnperez — CHANGES_REQUESTED** [review]

File: pkg/asset/installconfig/powervs/platform.go (hunk 6 above)
Comment (23TNC) [reply pkg/asset/installconfig/powervs/platform.go]: A few meetings ago, I believe you said that all of the os.Getenv and survey questions were located in session.go.  When I raised that platform.go had a bunch you seemed confused but accepted it.  As this is the last variable in this file that required the os package for os.Getenv I just moved it to session.go along with the others.

I suppose I also don't understand what version of session you are referring to above.  As the variables are coming out of the users current environment, I read session as the current environment.  If I pulled another shell I'd get another environment with different variables.  I guess I was using environment and session interchangeably.

I am guessing session in this case refers to an ibmcloud session or a powervs session or something.  Probably an ibmpisession object... which begs the question why aren't we returning just the ibmpisession object in that case, and why we would be ok carrying the Creds object along side the session object.

Idk... I can put this variable wherever.

File: pkg/asset/installconfig/powervs/session.go (hunk 5 above)
Comment (23TNC) [reply pkg/asset/installconfig/powervs/session.go:241]: I never did well in English I'm afraid.  I just wrote this out as if I was writing a description of a variable for function documentation,  Usually something like "var: The thing the var does or is"  Should I just complete the sentence?  "This is the"  Or would you prefer I just drop the period?

File: pkg/asset/installconfig/powervs/session.go (hunk 3 above)
Comment (23TNC) [reply pkg/asset/installconfig/powervs/session.go:142]: That comes down in pull request 51.  I'd actually prefer to solve this by making all of these use the same method, getEnv, and to add a bit of functionality to getEnv, or wrap getEnv... because all of these variables are handled the same way, and require survey prompts, and error checks... I figure we could drop that all into one function we can just call.  I figure we could add support for checking files for the same information at that time too.

Basically, 
findStuff( ["VarName, VarName2"], "Survey Prompt", "Survey Description", "FileName" )
Check if any of the permutations of the variable name have been exported with os.Getenv
If not, check if any of the permutations of the variable name exist in the provided file name (if provided)
If not, prompt the user with the survey prompt, and description

But this change was out of scope for this PR so I just moved this over.

File: pkg/asset/installconfig/powervs/platform.go (hunk 6 above)
Comment (23TNC) [reply pkg/asset/installconfig/powervs/platform.go]: I figure if we are keeping os image override in platform.go we can move all the region/zone stuff to region.go and call it from session.go.  Would that make more sense?

File: pkg/asset/installconfig/powervs/session.go (hunk 5 above)
Comment (clnperez) [reply pkg/asset/installconfig/powervs/session.go:241]: I was asking about the other providers. I've seen projects not like punctuation, but I did a quick grep to see and we should always add them, it looks like. (grep -r "Help:" pkg/asset/installconfig/*)

File: pkg/asset/installconfig/powervs/session.go (hunk 3 above)
Comment (clnperez) [reply pkg/asset/installconfig/powervs/session.go:142]: Sure, if we can find a way to streamline these later that's fine. But okay with deleting this check for now?

File: pkg/asset/installconfig/powervs/platform.go (hunk 6 above)
Comment (clnperez) [reply pkg/asset/installconfig/powervs/platform.go]: > I read session as the current environment.

Now I see why these have all been put into that struct. That wasn't my use of this.

>Probably an ibmpisession object.

Yep

> why aren't we returning just the ibmpisession object in that case,

I can't remember why I didn't do this. Something about it didn't work, but I agree that it does seem to make more sense on paper to return it instead.

> I figure if we are keeping os image override in platform.go we can move all the region/zone stuff to region.go and call it from session.go. Would that make more sense?

Yah that sounds good

**Comment by 23TNC** [comment]
Closing this due to add-ipi rebase, opened PR 58 that replaces this
